exchanges/gateio/ws_ob_update_manager.go

```diff
@@ -35,172 +44,217 @@ type pendingUpdate struct {
+	select {
+	case <-ctx.Done():
+		c.clearWithLock()
+		return ctx.Err()
+	case <-time.After(delay):
+	}
+
+	// Setting deadline to error out instead of waiting for rate limiter delay which excessively builds a backlog of
+	// pending updates.
+	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(deadline))
+	defer cancel()
+
+	book, err := e.fetchOrderbook(ctx, pair, a, limit)
+	if err != nil {
+		c.clearWithLock()
+		return err
+	}
+
+	if err := c.waitForUpdate(ctx, book.LastUpdateID+1); err != nil {
+		c.clearWithLock()
+		return err
+	}
+
+	c.mtx.Lock() // Lock here to prevent ws handle data interference with REST request above.
+	defer func() {
+		c.clearNoLock()
+		c.mtx.Unlock()
+	}()
+
+	if err := e.Websocket.Orderbook.LoadSnapshot(book); err != nil {
+		return err
+	}
+
+	return c.applyPendingUpdates(e)
+}
+
+// TODO: When subscription config is added for all assets update limits to use sub.Levels
+func (c *updateCache) extractOrderbookLimit(e *Exchange, a asset.Item) (uint64, error) {
 	switch a {
 	case asset.Spot:
-		sub := g.Websocket.GetSubscription(spotOrderbookUpdateKey)
+		sub := e.Websocket.GetSubscription(spotOrderbookUpdateKey)
 		if sub == nil {
-			return fmt.Errorf("no subscription found for %q", spotOrderbookUpdateKey)
+			return 0, fmt.Errorf("%w for %q", subscription.ErrNotFound, spotOrderbookUpdateKey)
 		}
-		// There is no way to set levels when we subscribe for this specific subscription case.
+		// There is no way to set levels when we subscribe for this specific channel
 		// Extract limit from interval e.g. 20ms == 20 limit book and 100ms == 100 limit book.
-		limit = uint64(sub.Interval.Duration().Milliseconds()) //nolint:gosec // No overflow risk
-	case asset.USDTMarginedFutures, asset.USDCMarginedFutures:
-		limit = futuresOrderbookUpdateLimit
+		lim := uint64(sub.Interval.Duration().Milliseconds()) //nolint:gosec // No overflow risk
+		if lim != 20 && lim != 100 {
+			return 0, fmt.Errorf("%w: %d. Valid limits are 20 and 100", errInvalidOrderbookUpdateInterval, lim)
+		}
+		return lim, nil
+	case asset.USDTMarginedFutures, asset.CoinMarginedFutures:
+		return futuresOrderbookUpdateLimit, nil
 	case asset.DeliveryFutures:
-		limit = deliveryFuturesUpdateLimit
+		return deliveryFuturesUpdateLimit, nil
 	case asset.Options:
-		limit = optionOrderbookUpdateLimit
+		return optionOrderbookUpdateLimit, nil
+	default:
+		return 0, fmt.Errorf("%w: %q", asset.ErrNotSupported, a)
 	}
+}
 
-	book, err := g.UpdateOrderbookWithLimit(ctx, pair, a, limit)
-
-	c.mtx.Lock() // lock here to prevent ws handle data interference with REST request above
-	defer func() {
-		c.updates = nil
-		c.updating = false
-		c.mtx.Unlock()
-	}()
+// waitForUpdate waits for an update with an ID >= nextUpdateID
+func (c *updateCache) waitForUpdate(ctx context.Context, nextUpdateID int64) error {
+	var updateListLastUpdateID int64
+	c.mtx.Lock()
+	updateListLastUpdateID = c.updates[len(c.updates)-1].update.UpdateID
```

Comment:
```suggestion
	updateListLastUpdateID := c.updates[len(c.updates)-1].update.UpdateID
```
and remove the declaration above

Reply:
🔥